wd_helper.au3

```diff
@@ -2196,6 +2196,52 @@ Func _WD_CheckContext($sSession, $bReconnect = Default, $vTarget = Default)
 	Return SetError(__WD_Error($sFuncName, ($iResult) ? $_WD_ERROR_Success : $_WD_ERROR_Exception), 0, $iResult)
 EndFunc   ;==>_WD_CheckContext
 
+; #FUNCTION# ====================================================================================================================
+; Name ..........: _WD_FindElement_ByRegExp
+; Description ...: Find element by matching attributes values using regular expression
+; Syntax ........: _WD_FindElement_ByRegExp($sSession, $sMode, $sRegEx[, $sRegExFlags = ""[, $bAll = False]])
+; Parameters ....: $sSession            - Session ID from _WD_CreateSession
+;                  $sMode               - Attribute of the element which should be matched, e.g. `id`, `style`, `class` etc.
+;                  $sRegEx              - RegEx in JavaScript format
+;                  $sRegExFlags         - [optional] RegEx Flags. Default is "".
+;                  $bAll                - [optional] Fetch a list of all elements that fits to the RegEx pattern ? Default is False - return only first element
+; Return values .: Success - Matched element
+;                  Failure - Response from _WD_ExecuteScript() and sets @error to value returned from _WD_ExecuteScript()
+; Author ........: TheDcoder
+; Modified ......: mLipok
+; Remarks .......: The RegEx matching is done by the browser's JavaScript engine so AutoIt's RegEx rules may not accurately work
+;                  in this function. You may refer to this cheatsheet for further information:
+;                  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Cheatsheet
+;                  The matching elements are returned in the same format as the results from _WD_FindElement
+;                  As JavaScript's RegEx engine differs from AutoIt's PCRE engine you can use https://regex101.com with FLAVOR set to: ECMAScript (JavaScript) to validate your RegEx
+; Related .......:
+; Link ..........:
+; Example .......: _WD_FindElement_ByRegExp($sSession, 'class', 'button-[0-9]', 'i', True)
+; ===============================================================================================================================
+Func _WD_FindElement_ByRegExp($sSession, $sMode, $sRegEx, $sRegExFlags = "", $bAll = False)
```

Comment:
using `$bAll = False` like this:
```autoit
_WD_FindElement_ByRegExp($sSesion, 'data-url', '\/danp2.au3webdr.*\/commits_list_item', 'i', False)
```
it still returns all 35 elements 
should only return first as string not as array

Reply:
This fix [return string or array fix](https://github.com/Danp2/au3WebDriver/pull/301/commits/6ab8649f0f227dfc82d5b7bd730318e47a9a3eaf)

When I use this example:

```autoit
Func _EXAMPLE__WD_FindElement_ByRegExp($sSesion)
	_WD_Navigate($sSesion, 'https://github.com/Danp2/au3WebDriver/commits/master')
	_WD_LoadWait($sSesion)
;~ 	data-url="/Danp2/au3WebDriver/commits/8bdf053f3b97d0b585b9f5bf2d6607a67eb5a545/commits_list_item"

	Local $vResult = _WD_FindElement_ByRegExp($sSesion, 'data-url', '\/danp2.au3webdr.*\/commits_list_item', 'i', False)
	If @error Then ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & _
			" : _WD_FindElement_ByRegExp" & @CRLF)

	MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information", $vResult)

	Local $vResult = _WD_FindElement_ByRegExp($sSesion, 'data-url', '\/danp2.au3webdr.*\/commits_list_item', 'i', True)
	If @error Then ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & _
			" : _WD_FindElement_ByRegExp" & @CRLF)

	_ArrayDisplay($vResult, '$vResult')
EndFunc   ;==>_EXAMPLE__WD_FindElement_ByRegExp
```

gives me such results:
![image](https://user-images.githubusercontent.com/11089482/169611330-411cebf1-ceaa-4b99-8637-2da56249d3de.png)


What you think about ?

Reply:
You were converting all Javascript parameters to strings. The correct fix IMO is to change this one line --

`Local $sArguments = StringFormat('"%s", "%s", "%s", %s', $sMode, $sRegEx, $sRegExFlags, StringLower($bAll))`

Notice that the last `%s` is no longer surrounded by double quotes. Now that variable will be seen as a boolean instead of a string.

Reply:
Thanks this should do the trick.
But I want to keep my current version for getting know if there was single or multiple elements returned by the given `$sRegExPattern`.
For this reason I add `$iRow` to be returned as `@extended`

Reply:
Please check my current [refactoring + standardarization + error handling](https://github.com/Danp2/au3WebDriver/pull/301/commits/3fdfa3f12e4615e45273058102b9f34672e8fadc)


```autoit
Func _EXAMPLE__WD_FindElement_ByRegExp($sSesion)
	_WD_Navigate($sSesion, 'https://github.com/Danp2/au3WebDriver/commits/master')
	_WD_LoadWait($sSesion)
;~ 	data-url="/Danp2/au3WebDriver/commits/8bdf053f3b97d0b585b9f5bf2d6607a67eb5a545/commits_list_item"

	Local $vResult, $sRegExpPattern = ''


	$sRegExpPattern = '\/danp2.au3webdr.*\/commits_list_item'
	$vResult = _WD_FindElement_ByRegExp($sSesion, 'data-url', $sRegExpPattern, 'i', False)
	MsgBox($MB_OK + $MB_TOPMOST, "Information", _
			"$vResult = " & $vResult & @CRLF & _
			"" & @CRLF & _
			"@error = " & @error & @CRLF & _
			"@extended = " & @extended & @CRLF & _
			"")

	$sRegExpPattern = '\/danp2.*\/commits_list_item'
	$vResult = _WD_FindElement_ByRegExp($sSesion, 'data-url', $sRegExpPattern, 'i', True)
	If @error Then ConsoleWrite("! ---> @error=" & @error & "  @extended=" & @extended & _
			" : _WD_FindElement_ByRegExp" & @CRLF)

	_ArrayDisplay($vResult, '$vResult')
EndFunc   ;==>_EXAMPLE__WD_FindElement_ByRegExp
```
It gives me:
![image](https://user-images.githubusercontent.com/11089482/169648655-75dc600c-1d84-48c0-adce-3d421bf9d077.png)

Reply:
> Thanks this should do the trick. But I want to keep my current version for getting know if there was single or multiple elements returned by the given `$sRegExPattern`. For this reason I add `$iRow` to be returned as `@extended`

In my opinion we should not alter this code because it will impact performance for cases where we don't need to know the number of matches. This can still be accomplished by the calling code by setting `$bAll` to true and counting the elements manually.

The number of matches can be a lot if a common `mode` attribute like class is chosen and the performance impact can be significant as the code will have to go through each element even when `$bAll` is not set.

In sort, I think we are over accommodating this function by sacrificing performance, just my two cents :smile:

Reply:
I agree with @TheDcoder. The Javascript code should be reverted to the earlier incarnation and then fix the parameter passing so that the $bAll functionality works properly. @mlipok Unless I misunderstood your comment, you should still be able to get the count of matching elements.

Reply:
Ok.
Yes.

Reply:
See a7383967fc813f3012b7a6b1739ae82aa7836506 for my attempt to address this.

Reply:
> See [a738396](https://github.com/Danp2/au3WebDriver/commit/a7383967fc813f3012b7a6b1739ae82aa7836506) for my attempt to address this.

I see you have reverte this by: [Revert to earlier version of Javascript](https://github.com/Danp2/au3WebDriver/pull/301/commits/a7383967fc813f3012b7a6b1739ae82aa7836506)

Ok.